File: rip_quick_add.py
from datetime import datetime, timedelta
from time import mktime
from typing import Optional, Tuple

import parsedatetime as pdt
from requests import Request


cal = pdt.Calendar()
PDT_FAIL = 0
PDT_DATE = 1

def extract_datetime(text: str) -> Tuple[datetime, int]:
    result = cal.parse(text)
    if result[1] == PDT_FAIL:
        raise ValueError('could not parse timestamp from event')
    dt = datetime.fromtimestamp(mktime(result[0]))
    return dt, result[1]


def dt_format(dt: datetime, flag: int) -> str:
    if flag == PDT_DATE:
        return dt.strftime('%Y%m%d')
    return dt.strftime('%Y%m%dT%H%M%S')


def create_invite_url(text: str) -> Optional[str]:
    try:
        start, flag = extract_datetime(text)
    except ValueError:
        return None

    if flag == PDT_DATE:
        end = start + timedelta(days=1)
    else:
        end = start + timedelta(hours=1)

    dates = '/'.join(dt_format(t, flag) for t in [start, end])

    params = {
        'action': 'TEMPLATE',
        'text': text,
        'dates': dates,
    }
    p = Request('GET', 'https://www.google.com/calendar/render',
                params=params).prepare()
    return p.url


if __name__ == '__main__':
    text = 'Dinner with Michael 7 p.m. tomorrow'
    url = create_invite_url(text)
    print(url)

File: rip_quick_add.js
import * as chrono from 'chrono-node';
import dayjs from 'dayjs';

function dateRange(start, end, isAllDay) {
  let formatString = 'YYYYMMDD[T]HHmmss';

  if (isAllDay) {
    formatString = 'YYYYMMDD';
  }

  return [start, end].map(t => t.format(formatString)).join('/');
}

function parse(text) {
  if (!text) {
    throw new Error('invalid input text');
  }

  const results = chrono.parse(text);

  if (results.length === 0) {
    throw new Error('could not find time data');
  }

  const result = results[0];

  const eventTitle = text.replace(result.text, "").trim();
  const startDate = result.start;
  const isAllDay = !startDate.isCertain('hour');
  const start = dayjs(startDate.date());
  let end = null;

  if (!start.isValid()) {
    throw new Error('could not find time data');
  }

  if (result.end !== null) {
    end = dayjs(result.end.date());
  }

  if (end === null) {
    if (isAllDay) {
      end = start.add(1, 'days');
    } else {
      end = start.add(1, 'hours');
    }
  }

  const dates = dateRange(start, end, isAllDay);
  return { text: eventTitle, dates }
}

// TODO: Update this to the latest version.
export function createEventUrl(text) {
  let data;
  try {
    data = parse(text);
    console.log(data);
  } catch (err) {
    console.log(err);
    return null;
  }

  data.action = 'TEMPLATE';

  const baseUrl = 'https://www.google.com/calendar/render';
  const params = new URLSearchParams(data);
  return `${baseUrl}?${params}`;
}
